docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,388 @@
+can run on (this is a common requirement for iSCSI, FibreChannel, or NFS clusters).
+
+Because all nodes in a Kubernetes cluster possess a special node name label, it would
+be possible for a controller to observe the scheduling decision of a pod using an
+unsafe volume and "attach" that volume to the node, and also observe the deletion of
+the pod and "detach" the volume from the node. The node would then require that these
+unsafe volumes be "attached" before allowing pod execution. Attach and detach may
+be recorded on the PVC or PV as a new field or materialized via the selection labels.
+
+Possible sequence of operations:
+
+1. Cluster administrator creates a RWO iSCSI persistent volume, available only to
+   nodes with the label selector `storagecluster=iscsi-1`
+2. User requests an RWO volume and is bound to the iSCSI volume
+3. The user creates a pod referencing the PVC
+4. The scheduler observes the pod must schedule on nodes with `storagecluster=iscsi-1`
+   (alternatively this could be enforced in admission) and binds to node `A`
+5. The kubelet on node `A` observes the pod references a PVC that specifies RWO which
+   requires "attach" to be successful
+6. The attach/detach controller observes that a pod has been bound with a PVC that
+   requires "attach", and attempts to execute a CAS update on the PVC/PV attaching
+   it to node `A` and pod 1
+7. The kubelet observes the attach of the PVC/PV and executes the pod
+8. The user terminates the pod
+9. The user creates a new pod that references the PVC
+10. The scheduler binds this new pod to node `B`, which also has `storagecluster=iscsi-1`
+11. The kubelet on node `B` observes the new pod, but sees that the PVC/PV is bound
+    to node `A` and so must wait for detach
+12. The kubelet on node `A` completes the deletion of pod 1
+13. The attach/detach controller observes the first pod has been deleted and that the
+    previous attach of the volume to pod 1 is no longer valid - it performs a CAS
+    update on the PVC/PV clearing its attach state.
+14. The attach/detach controller observes the second pod has been scheduled and
+    attaches it to node `B` and pod 2
+15. The kubelet on node `B` observes the attach and allows the pod to execute.
+
+If a partition occurred after step 11, the attach controller would block waiting
+for the pod to be deleted, and prevent node `B` from launching the second pod.
+The fencing controller, upon observing the partition, could signal the iSCSI servers
+to firewall node `A`. Once that firewall is in place, the fencing controller could
+break the PVC/PV attach to node `A`, allowing steps 13 onwards to continue.
+
+
+### User interface changes
+
+Clients today may assume that force deletions are safe. We must appropriately
+audit clients to identify this behavior and improve the messages. For instance,
+`kubectl delete --grace-period=0` could print a warning and require `--confirm`:
```

Comment:
If the pod has pending finalizers, --grace-period=0 will not delete the pod from the key-value store. Rather than overloadin the `grace-period` flag, can we define a new flag, like `--forceful-deletion`, which will instruct the API server to also ignore pending finalizers?